### PR TITLE
Creation of PHAR using Nix, in a reproducible environment

### PR DESCRIPTION
**Pull request by drupol**
Hey there,

I saw the message from @sebastianbergmann on Mastodon (https://main.elk.zone/phpc.social/@sebastian/111528636893710027) and I thought... How about using Nix to do it?

This PR shows how it could be done using 2 different approaches:

1. By using the existing `build-phar.sh` and _nixifying_ it, providing `box` without needing to install it with Composer and build the PHAR in a non isolated environment.
2. By using a `flake.nix` file
  - Run `nix build .#build-phar-script` to build the script to build the PHAR (result in `result/` directory)
  - Run `nix build .#phar` to build the PHAR in isolation (result in `result/` directory)
  - Run `nix run github:drupol/BackwardCompatibilityCheck/creation-of-phar-using-nix` to immediately run the app 
    from your workstation without installing it, in total isolation.
  - Run `nix develop` to create a development shell containing PHP, Composer and Box and the `build-phar-script`

Both methods work well, and are totally reproducible. My favourite approach is to use the second one which is more flexible and versatile, but anyway one or the other, the environment building the PHAR is the same... (pun intended!). Note, the PHAR is not reproducible, `box` doesn't produce reproducible PHAR files by default, unfortunately (read how to make it reproducible [here](https://github.com/box-project/box/issues/447#issuecomment-1075102743))

I'm pretty sure this PR won't be merged, but I just wanted to see if this was doable, I hope this is going to give ideas to some PHP community members :)

It took me 25 minutes to do, mostly copy-pasting things here and there.

### CODE REVIEW

File: flake.lock
Comment (Ocramius) [inline flake.lock:1]: Overall OK with having a `flake.lock`, but we really should have Renovate update this :thinking:

File: .github/workflows/release-phar.yml
Comment (Ocramius) [inline .github/workflows/release-phar.yml]: Is there a pinnable version for this?

Comment (drupol) [reply .github/workflows/release-phar.yml]: Sure, I'll do it, I'm busy fixing something inside the flake!

Comment (drupol) [reply .github/workflows/release-phar.yml]: OK, everything fixed, I was missing the `installPhase` ! Dumb me.

File: flake.lock
Comment (drupol) [reply flake.lock:1]: It supports it now: https://github.com/renovatebot/renovate/issues/8000

But I don't remember if it updates all the inputs or only `nixpkgs` inputs.

File: .github/workflows/release-phar.yml
Comment (Ocramius) [reply .github/workflows/release-phar.yml]: Yes, but I'm still wondering if we can avoid referencing `@main` :D

Comment (drupol) [reply .github/workflows/release-phar.yml]: Yes, I understood, it's already removed :)

File: flake.lock
Comment (Ocramius) [reply flake.lock:1]: Right, can we somehow enable it in our config? Or is it the default behavior? :thinking: 

In that thread, I [see](https://github.com/renovatebot/renovate/issues/8000#issuecomment-1312289641)

```json
"nix": {
  "enabled": true
},
```

File: .gitattributes
Comment (Ocramius) [inline .gitattributes:8]: Weird that this was ignored: good catch!

File: flake.lock
Comment (drupol) [reply flake.lock:1]: Apparently, it is updating all the inputs! That's great!

https://github.com/loophp/php-src-nix/pull/52

File: flake.nix
Comment (Ocramius) [inline flake.nix]: This vendor hash: will it be updated automatically? I don't see it being in any CI setup.

Sorry BTW for reviewing this in hiccups: multitasking.

File: flake.nix
Comment (Ocramius) [inline flake.nix]: I think we want the version to not be declared at all: either inferred or use `dev` or such here, IMO.

File: flake.nix
Comment (Ocramius) [inline flake.nix]: I think we want the version to not be declared at all: either inferred or use `dev` or such here, IMO.

File: renovate.json
Comment (Ocramius) [inline renovate.json]: Luckily, this was already enabled in [`local>Ocramius/.github:renovate-config`](https://github.com/Ocramius/.github/blob/c77629415fd5a57ef23d9372075f406e7bf98403/renovate-config.json#L19-L24) :D

**Review by Ocramius — CHANGES_REQUESTED** [review]
Need some more clarity on the degree of manual intervention on `flake.nix`: can't really manually change its contents at every release

File: flake.nix
Comment (drupol) [reply flake.nix]: This hash should be changed every time you change the `composer.lock` file.

That said, I just pushed a commit and update my approach to the building of the PHAR.

File: flake.nix
Comment (drupol) [reply flake.nix]: Fixed, removed it.

File: flake.nix
Comment (drupol) [reply flake.nix]: Done.

File: renovate.json
Comment (drupol) [reply renovate.json]: Removed.

File: build-phar.sh
Comment (Ocramius) [inline build-phar.sh]: Eh, now we moved a random hash in here? 😬

Comment (drupol) [reply build-phar.sh]: Let me delete that script, it's not needed any more

Comment (drupol) [reply build-phar.sh]: It was needed in the first method, but since I switched to flake, it's not needed any more.

File: composer.json
Comment (Ocramius) [inline composer.json]: Really part of `require`? :thinking:

File: resign.php
Comment (Ocramius) [inline resign.php]: This date seems very arbitrary? :thinking:

File: scoper.inc.php
Comment (Ocramius) [inline scoper.inc.php]: I smell tabs :D not important, but noticed

File: composer.json
Comment (Ocramius) [reply composer.json]: Maybe `require-dev`?

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: composer.json
Comment (drupol) [reply composer.json]: True, will do tomorrow, went to bed already 😴

File: flake.nix
Comment (Ocramius) [inline flake.nix]: Ok, so we're back at having the hash here: how would one approach a patch by Renovate where `composer.lock` is being touched? :thinking:

Comment (drupol) [reply flake.nix]: Yeah... I can't basically get rid of it, I tried... and there's a reason to have it there.

Since this PR is quite messy, how about lively discuss this during an informal meeting at your best convenience?

Comment (Ocramius) [reply flake.nix]: > Since this PR is quite messy, how about lively discuss this during an informal meeting at your best convenience?

Can probably do next week about this.

Meanwhile, I have an example of a repo that has a `Cargo.lock` and no hash: https://github.com/Ocramius/oci-srm-server-mock-rust/blob/383e9bd1e99fca42578bd885ae01ff7c74b163ed/flake.nix

Perhaps `php.buildComposerProject` to be improved? Or are we missing sha256 for downloaded vendor projects as part of Composer itself? I see that `Cargo.lock` has [checksums](https://github.com/Ocramius/oci-srm-server-mock-rust/blob/383e9bd1e99fca42578bd885ae01ff7c74b163ed/Cargo.lock#L246), while [we don't](https://github.com/Roave/BackwardCompatibilityCheck/blob/9b45bc3431f3dc651a8f2d6e199ca0d1e60a9978/composer.lock#L21)

Comment (Ocramius) [reply flake.nix]: Note: could be worked around by assuming `--prefer-source` to download via `git`, where we at least have a SHA1

Comment (drupol) [reply flake.nix]: Let's do that next week then, ping me on slack, matrix or even telegram when you're ready.

Regarding the builder, there might be room for improvements for sure, and what you said is actually something I tried to fix in https://github.com/NixOS/nixpkgs/pull/225401

I wish I could have a mapping between the `composer.lock` file and the `vendorHash`... But I have the feeling that it would require to update Composer first. That's an interesting discussion for sure.

You can find the sources of the builder here: https://github.com/NixOS/nixpkgs/tree/master/pkgs/build-support/php

Feel free to suggest improvements, having you on board on this is definitely a big plus.

Comment (Ocramius) [reply flake.nix]: > But I have the feeling that it would require to update Composer first.

If we only do source-based installs, we can rely on `git` for some basic shasum, since Composer's is currently useless. 

For now, I'd be happy if **in this patch** the `vendorHash` could be generated inside the github action: that would be a sufficient approach to move forward, IMO

File: flake.nix
Comment (Ocramius) [inline flake.nix]: This is still a blocker for merging this PR: `composer.lock` changes **daily**, so this added maintenance burden here blocks merge

Comment (drupol) [reply flake.nix]: I know... there's no solution to this, sadly. And I haven't anything to propose to have it automatically updated.
There's this PR: https://github.com/Mic92/nix-update/pull/211 that once merged could be a potential solution, but as of today, nothing yet.

Comment (Ocramius) [reply flake.nix]: Yeah, just wanted to spell out the problem here 👍

**Review by Ocramius — APPROVED** [review]
:ship: thanks @drupol! 

I know this isn't exactly what you wanted, but I'm also glad that we found a middle-ground here, and that we introduced nix and nix-flakes in the entire dev stack.

Happy that the discussions/requirements led to more understanding of what package maintainers need from this end too :-) 

I'll now merge and try out the release script, thanks for the help!

**Comment by Ocramius** [comment]
Aw, still need to fix PHP 8.1 :D 

Should be trivial, though: I'm on it.

**Comment by Ocramius** [comment]
The CI failure is due to an ancient build of the require-checker `.phar`.

Unsure why we `wget` it, but not a problem for this patch to solve.

**Comment by Ocramius** [comment]
Well, this is merged, but no phar got tagged :D 

https://github.com/Roave/BackwardCompatibilityCheck/actions/workflows/release-phar.yml

I suspect the release being triggered by a bot prevents this task from being triggered: https://github.com/Roave/BackwardCompatibilityCheck/blob/4f6c882dce148eb28ede2adf47087ae47621d849/.github/workflows/release-phar.yml#L5-L8

**Comment by Ocramius** [comment]
The publishing payload, meanwhile, taken from https://api.github.com/repos/roave/BackwardCompatibilityCheck/events :

```json
  {
    "id": "34310185998",
    "type": "ReleaseEvent",
    "actor": {
      "id": 41898282,
      "login": "github-actions[bot]",
      "display_login": "github-actions",
      "gravatar_id": "",
      "url": "https://api.github.com/users/github-actions[bot]",
      "avatar_url": "https://avatars.githubusercontent.com/u/41898282?"
    },
    "repo": {
      "id": 108544542,
      "name": "Roave/BackwardCompatibilityCheck",
      "url": "https://api.github.com/repos/Roave/BackwardCompatibilityCheck"
    },
    "payload": {
      "action": "published",
      "release": {
        "url": "https://api.github.com/repos/Roave/BackwardCompatibilityCheck/releases/134622383",
        "assets_url": "https://api.github.com/repos/Roave/BackwardCompatibilityCheck/releases/134622383/assets",
        "upload_url": "https://uploads.github.com/repos/Roave/BackwardCompatibilityCheck/releases/134622383/assets{?name,label}",
        "html_url": "https://github.com/Roave/BackwardCompatibilityCheck/releases/tag/8.5.0",
        "id": 134622383,
        "author": {
          "login": "github-actions[bot]",
          "id": 41898282,
          "node_id": "MDM6Qm90NDE4OTgyODI=",
          "avatar_url": "https://avatars.githubusercontent.com/in/15368?v=4",
          "gravatar_id": "",
          "url": "https://api.github.com/users/github-actions%5Bbot%5D",
          "html_url": "https://github.com/apps/github-actions",
          "followers_url": "https://api.github.com/users/github-actions%5Bbot%5D/followers",
          "following_url": "https://api.github.com/users/github-actions%5Bbot%5D/following{/other_user}",
          "gists_url": "https://api.github.com/users/github-actions%5Bbot%5D/gists{/gist_id}",
          "starred_url": "https://api.github.com/users/github-actions%5Bbot%5D/starred{/owner}{/repo}",
          "subscriptions_url": "https://api.github.com/users/github-actions%5Bbot%5D/subscriptions",
          "organizations_url": "https://api.github.com/users/github-actions%5Bbot%5D/orgs",
          "repos_url": "https://api.github.com/users/github-actions%5Bbot%5D/repos",
          "events_url": "https://api.github.com/users/github-actions%5Bbot%5D/events{/privacy}",
          "received_events_url": "https://api.github.com/users/github-actions%5Bbot%5D/received_events",
          "type": "Bot",
          "site_admin": false
        },
        "node_id": "RE_kwDOBnhCHs4IBiyv",
        "tag_name": "8.5.0",
        "target_commitish": "8.5.x",
        "name": "8.5.0",
        "draft": false,
        "prerelease": false,
        "created_at": "2023-12-20T13:08:55Z",
        "published_at": "2023-12-20T13:08:57Z",
        "assets": [

        ],
        "tarball_url": "https://api.github.com/repos/Roave/BackwardCompatibilityCheck/tarball/8.5.0",
        "zipball_url": "https://api.github.com/repos/Roave/BackwardCompatibilityCheck/zipball/8.5.0",
        "body": "### Release Notes for [8.5.0](https://github.com/Roave/BackwardCompatibilityCheck/milestone/59)\n\nFeature release (minor)\n\n### 8.5.0\n\n- Total issues resolved: **0**\n- Total pull requests resolved: **2**\n- Total contributors: **2**\n\n#### enhancement\n\n - [755: Fix: Consistently indent with 2 spaces](https://github.com/Roave/BackwardCompatibilityCheck/pull/755) thanks to @localheinz\n - [754: Creation of PHAR using Nix, in a reproducible environment](https://github.com/Roave/BackwardCompatibilityCheck/pull/754) thanks to @drupol\n",
        "mentions_count": 2,
        "mentions": [
          {
            "avatar_url": "https://avatars.githubusercontent.com/u/252042?v=4",
            "login": "drupol",
            "profile_name": "Pol Dellaiera",
            "profile_url": "https://github.com/drupol",
            "avatar_user_actor": true
          },
          {
            "avatar_url": "https://avatars.githubusercontent.com/u/605483?v=4",
            "login": "localheinz",
            "profile_name": "Andreas Möller",
            "profile_url": "https://github.com/localheinz",
            "avatar_user_actor": true
          }
        ],
        "short_description_html": "<h3>Release Notes for <a href=\"https://github.com/Roave/BackwardCompatibilityCheck/milestone/59\">8.5.0</a>\n</h3>\n<p>Feature release (minor)</p>\n<h3>8.5.0</h3>\n<ul>\n<li>Total issues resolved: <strong>0</strong>\n</li>\n<li>Total pull requests resolved: <strong>2</strong>\n</li>\n<li>Total contributors: <strong>2</strong>\n</li>\n</ul>\n<h4>enhancement</h4>\n<ul>\n<li>\n<a href=\"https://github.com/Roave/BackwardCompatibilityCheck/pull/755\" data-hovercard-type=\"pull_request\" data-hovercard-url=\"/Roave/BackwardCompatibilityCheck/pull/755/hovercard\">755: Fix: Consistently indent with 2 spaces</a> thank…</li>\n</ul>",
        "is_short_description_html_truncated": true
      }
    },
    "public": true,
    "created_at": "2023-12-20T13:08:57Z",
    "org": {
      "id": 3029050,
      "login": "Roave",
      "gravatar_id": "",
      "url": "https://api.github.com/orgs/Roave",
      "avatar_url": "https://avatars.githubusercontent.com/u/3029050?"
    }
  },
```

**Comment by drupol** [comment]
Let me have a look at the workflow at the end of the day

**Comment by Ocramius** [comment]
Identified that I need `ORGANIZATION_ADMIN_TOKEN` here, roughly: https://github.com/Roave/BackwardCompatibilityCheck/blob/4f6c882dce148eb28ede2adf47087ae47621d849/.github/workflows/release-on-milestone-closed.yml#L24

**Comment by Ocramius** [comment]
Now it did run, but failed :D 

```
/home/runner/work/_temp/8083e12c-ba48-4ca7-b23b-3bba47cad80a.sh: line 1: nix run .#build-phar-script: command not found
```

https://github.com/Roave/BackwardCompatibilityCheck/actions/runs/7276447892

That's... curious

**Pull request merged**
